dwave_networkx/algorithms/matching.py

```diff
@@ -13,10 +13,164 @@
+        bqm = dimod.BinaryQuadraticModel.empty('BINARY')
+
+    # add the edges of G as variables
+    for edge in G.edges:
+        bqm.add_variable(frozenset(edge), 0)
+
+    for node in G:
+        for edge0, edge1 in itertools.combinations(G.edges(node), 2):
+            u = frozenset(edge0)
+            v = frozenset(edge1)
+            bqm.add_interaction(u, v, 1)
+
+    return bqm
+
+
+def maximal_matching_bqm(G, lagrange=None):
+    """Return a BQM with ground states corresponding to the possible maximal
+    matchings of G.
+
+    A matching is a subset of edges in which no node occurs more than
+    once. A maximal matching is one in which no edges from G can be
+    added without violating the matching rule.
+
+    Finding maximal matchings can be done is polynomial time, so this BQM
+    is only efficient when combined with other constraints and objectives.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        The graph on which to find a maximal matching.
+
+    lagrange : float (optional)
+        The Lagrange multiplier for the matching constraint. Should be postivie
+        and greater than `max_degree - 2`.
+        Defaults to `1.25 * (max_degree - 2)`.
+
+    Returns
+    -------
+    bqm : :class:`dimod.BinaryQuadraticModel`
+        A binary quadratic model with ground states corresponding to a maximal
+        matching. The variables of the BQM are the edges of `G` as frozensets.
+        The BQM's ground state energy is 0 by construction.
+
+    """
+    bqm = matching_bqm(G)
+
+    if lagrange is None:
+        delta = max((G.degree[v] for v in G), default=0)
+        lagrange = max(1.25 * (delta - 2), 1)
+
+    bqm.scale(lagrange)
+
+    for node0, node1 in G.edges:
+        # (1 - y_v - y_u + y_v*y_u) <- see paper
+
+        bqm.offset += 1
+
+        for edge in G.edges(node0):
+            bqm.linear[frozenset(edge)] -= 1
+
+        for edge in G.edges(node1):
+            bqm.linear[frozenset(edge)] -= 1
+
+        for edge0 in G.edges(node0):
+            u = frozenset(edge0)
+            for edge1 in G.edges(node1):
+                v = frozenset(edge1)
+                if u == v:
+                    bqm.linear[u] += 1
+                else:
+                    bqm.add_interaction(u, v, 1)
+
+    return bqm
+
+
+def min_maximal_matching_bqm(G, maximal_lagrange=2, matching_lagrange=None):
+    """Return a BQM with ground states corresponding to a minimum maximal
+    matching.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        The graph on which to find a minimum maximal matching.
+
+    maximal_lagrange : float (optional, default=2)
+        The Lagrange multiplier for the maximal constraint. Should be greater
+        than 1.
+
+    matching_lagrange : float (optional)
+        The Lagrange multiplier for the matching constraint. Should be postivie
+        and greater than `maximal_lagrange * max_degree - 2`.
+        Defaults to `1.25 * (maximal_lagrange * max_degree - 2)`.
```

Comment:
Here we're referencing a default set in a different function. If the default changes there, there's a high risk this remains stale. Better to reference the function.